### PR TITLE
Process link clear

### PR DESCRIPTION
**Pull request by rht**
TODO: add a gc to a process waitfors

### CODE REVIEW

**Comment by jbenet** [comment]
looking good. i think this is right (because no traversal is needed afterwards).

one option may be to make `Clear()` remove the process link from the lists.

**Comment by rht** [comment]
Undid the `ParentClear`, `ClearChild` -> `Clear` commit. Because this will have to add extra check whether parent/child is nil, or pl is cleared. Can't tell if this will simplify/complicate the code.

**Comment by rht** [comment]
@jbenet RFCR

File: impl-mutex.go [hunk 1]
```diff
@@ -94,7 +101,14 @@ func (p *process) AddChild(child Process) {
...
 	p.waitfors = append(p.waitfors, pl)
 	p.children = append(p.children, pl)
+	indexOf := len(p.waitfors)
 	p.Unlock()
+	select {
+	case <-pl.cleared:
+		// remove processLink (gc)
+		p.waitfors[indexOf] = nil
+	default:
+	}
```
Comment (jbenet) [inline impl-mutex.go]: curious why it's here-- was just created?

**Comment by jbenet** [comment]
im not sure if this is correct-- i've to think about it more. 

maybe we should make a test that allocates tons of processes (some with `goprocess.Background()`) and then closes them all, and then accounts for memory having gc-ed correctly. thoughts?

**Comment by rht** [comment]
Indeed,

before:
initial: 282k, n-processes: 6255k, after: 7781k

this pr:
initial: 290k, n-processes: 7282k, after: 10045k

test:
`
func TestMemoryLeak(t *testing.T) {
	// Read initial memory stats
	initMemStats := new(runtime.MemStats)
	runtime.ReadMemStats(initMemStats)

	// Spawn processes
	var ps []Process
	for i := 0; i < 10000; i++ {
		p := WithParent(Background())
		ps = append(ps, p)
	}
	nMemStats := new(runtime.MemStats)
	runtime.ReadMemStats(nMemStats)

	// Close all processes
	for _, p := range ps {
		p.Close()
	}
	afterMemStats := new(runtime.MemStats)
	runtime.ReadMemStats(afterMemStats)
	fmt.Printf("initial: %dk, n-processes: %dk, after: %dk\n", initMemStats.HeapAlloc/1000, nMemStats.HeapAlloc/1000, afterMemStats.HeapAlloc/1000)
}
`

**Comment by jbenet** [comment]
> before:
> initial: 282k, n-processes: 6255k, after: 7781k
>
> this pr:
> initial: 290k, n-processes: 7282k, after: 10045k

does this mean this PR makes it worse?

**Comment by rht** [comment]
Just corrected the number (I forgot to set `ps` to `nil`), it is 8k, but yes still worse because each process link now has the cleared chan.

File: impl-mutex.go (hunk 1 above)
Comment (rht) [inline impl-mutex.go]: To unset the pl from waitfors. Can't think of a better way.

**Comment by jbenet** [comment]
could you try with:

```go
func TestMemoryLeak(t *testing.T) {
    var stats []*runtime.MemStats
    iters := 10
    fanout := 50

    measure := func() {
        s := new(runtime.MemStats)
        runtime.ReadMemStats(s)
        stats = append(stats, s)
    }

    spawn := func(ps []Process) {
        // Spawn processes
        for i := 0; i < fanout; i++ {
            p := WithParent(Background())
            ps = append(ps, p)

            for i := 0; i < fanout; i++ {
                p2 := WithParent(p)
                ps = append(ps, p2)

                for i := 0; i < fanout; i++ {
                    p3 := WithParent(p2)
                    ps = append(ps, p3)
                }
            }
        }
    }

    // Read initial memory stats
    for i := 0; i < iters; i++ {
        var ps []Process

        spawn(ps)
        measure() // read after alloc 

        // Close all processes
        for _, p := range ps {
            p.Close()
        }

        measure() // read after dealloc, but before gc

        // give some time for GC to act? could be smaller, dont know.
        <-time.After(time.Second) 
        measure() // read after gc
    }

    ps = nil

    for _, s := range stats {
        fmt.Printf("%dk, S.HeapAlloc/1000)
    }
}
```

**Comment by rht** [comment]
`<-time.After(time.Second)` could be replaced with explicit GC call.

No it is monotonically increasing
```
18258k
18262k
18295k
32818k
32822k
32857k
49642k
49646k
49685k
60940k
60944k
61013k
77708k
77712k
77756k
94409k
94413k
94463k
111115k
111119k
111169k
118873k
118877k
118939k
135592k
135596k
135645k
152297k
152301k
152351k
```

**Comment by jbenet** [comment]
> `<-time.After(time.Second)` could be replaced with explicit GC call.

ah yep, 

> No it is monotonically increasing

even with the explicit GC call?

**Comment by rht** [comment]
Yes even with explicit GC.

Here is annotated list:
```alloc: 17014k
dealloc: 17018k
gc: 16504k
alloc: 31995k
dealloc: 31999k
gc: 31459k
alloc: 48166k
dealloc: 48170k
gc: 45432k
alloc: 63041k
dealloc: 63045k
gc: 60298k
alloc: 76970k
dealloc: 76974k
gc: 74229k
alloc: 91289k
dealloc: 91293k
gc: 88546k
alloc: 105224k
dealloc: 105228k
gc: 102482k
alloc: 119260k
dealloc: 119265k
gc: 116514k
alloc: 133610k
dealloc: 133614k
gc: 130869k
alloc: 147542k
dealloc: 147546k
gc: 144800k
```

**Comment by rht** [comment]
Also looks like 1s sleep doesn't gc much, e.g.
```
alloc: 18258k
dealloc: 18262k
gc: 18295k
alloc: 32761k
dealloc: 32765k
gc: 32806k
alloc: 49448k
dealloc: 49452k
gc: 49513k
```

**Comment by whyrusleeping** [comment]
@rht fix this and I owe you a beer (or beverage of choice) the goprocess memory leak keeps killing my nodes.

**Comment by rht** [comment]
@whyrusleeping ...Will fix as soon as I can (now that the err's location has been pinpointed, unlike the get progressbar EOF case).

Was reminded of the relation: GC + runtime optimization ~ dreaming ~ neural net pruning.

**Comment by whyrusleeping** [comment]
:+1: :smile:

**Comment by rht** [comment]
Should have updated this earlier. The leak is very specific to the example. Basically since `[]Process` is not a pointer and so the slice doesn't get updated.

Of course there is still `Background()` leak.
This is fixed with replacing slice with map to keep track of the processes.

**Comment by rht** [comment]
Before processLink GC:
```
Memory growth after 2000 iteration with each 8000 processes: 86.02% after 3067k
PASS
ok  	_/home/rht/github/goprocess	71.881s
```

After GC commit:
```
Memory growth after 2000 iteration with each 8000 processes: 3.90% after 2853k
PASS
ok  	_/home/rht/github/goprocess	80.850s
```

**Comment by rht** [comment]
RFCR. It appears that goroutines are more expensive than just keeping track of child processes through `process.children`.

**Comment by jbenet** [comment]
@rht gj with the reduction! not fully fixed yet, but an improvement. take a look at the comment above o/

**Comment by rht** [comment]
ic, the classic pointer trick does make it reach plateau faster.

```
Memory growth after 2000 iteration with each 8000 processes: 0.29% after 2934k
PASS
ok  	_/home/rht/github/goprocess	79.492s
```

```
Memory growth after 2000 iteration with each 8000 processes: 0.28% after 2959k
PASS
ok  	_/home/rht/github/goprocess	82.493s
```

**Comment by rht** [comment]
The 2k iteration leak test is too long to fit into the test.

**Comment by jbenet** [comment]
sounds good, thanks!

**Pull request merged**